File: src/main/java/linkedlist/MyLinkedList.java
package linkedlist;

import org.jetbrains.annotations.NotNull;

import java.util.Iterator;
import java.util.Spliterator;
import java.util.function.Consumer;

public class MyLinkedList implements GLinkedList<Integer>, Iterable<Integer> {

  private Node headNode;

  // ------- Constructor -------

  public MyLinkedList(Integer... ints) {
    for (int i = (ints.length - 1); i >= 0; i--) {
      addFirst(ints[i]);
    }
  }


  // ------- Add -------

  public void addLast(Integer data) {
    Node newNode = new Node(data);

    // If list is null, make the headNode the newNode. Otherwise, find the end and add the new node
    if (headNode == null) {
      headNode = newNode;

    } else {
      Node currentNode  = headNode;

      // Loop to the end of the current list and link the last node to newNode
      while (currentNode.getNext() != null) {
        currentNode = currentNode.getNext();
      }
      currentNode.setNext(newNode);
    }
  }

  public void addAtIndex(Integer data, int index) {
    Node newNode = new Node(data);

    // If the index is 0, add a new head. Otherwise, find the index and add new Node.
    if (index == 0) {
      // If the list is null, create a new head, otherwise copy new head info
      if (headNode == null) {
        headNode = newNode;
        return;
      }

      // Copy info from headNode into new Node for second index
      Node newSecondNode = new Node(headNode.getData());
      newSecondNode.setNext(headNode.getNext());

      // Create a new Node from the input data and make it the first Node
      newNode.setNext(newSecondNode);
      headNode = newNode;

    } else {
      Node currentNode = headNode;
      int positionCounter = 0;

      // Iterate until the end of the list
      while (currentNode.getNext() != null) {
        if (positionCounter == index - 1) {
          newNode.setNext(currentNode.getNext());
          currentNode.setNext(newNode);
          return;
        }

        positionCounter += 1;
        currentNode = currentNode.getNext();
      }

      // Index is either the last (currently null) position, or out of bounds.
      if (positionCounter == index - 1) {
        currentNode.setNext(newNode);

      } else {
        throw new IndexOutOfBoundsException();
      }
    }
  }

  public void addFirst(Integer data) {
    addAtIndex(data, 0);
  }


  // ------- Set -------

  public void setLast(Integer data) {
    nullListCheck();

    Node currentNode = headNode;

    while (currentNode.getNext() != null) {
      currentNode = currentNode.getNext();
    }

    currentNode.setData(data);
  }

  public void setAtIndex(Integer data, int index) {
    nullListCheck();

    if (index == 0) {
      setFirst(data);
      return;
    }

    Node currentNode = headNode;
    int currentIndex = 0;

    while (currentNode != null && currentIndex <= index) {
      if (currentIndex == index) {
        currentNode.setData(data);
        return;
      }
      currentNode = currentNode.getNext();
      currentIndex += 1;
    }

    throw new IndexOutOfBoundsException();
  }

  public void setFirst(Integer data) {
    nullListCheck();
    headNode.setData(data);
  }


  // ------- Delete -------

  public void remove(Integer data) {
    nullListCheck();

    boolean dataDeleted = false;

    if (headNode.getData() == data) {
      removeFirst();
      dataDeleted = true;
    }

    Node currentNode = headNode;

    do {
      if (currentNode.getNext().getData() == data) {
        // Get the Node to delete (the next one) and it's reference (the one after).
        Node nodeToDelete = currentNode.getNext();
        Node nodeToDeleteLinkedNode = nodeToDelete.getNext();

        // Remove link from nodeToDelete
        nodeToDelete.setNext(null);

        // Fix the chain
        currentNode.setNext(nodeToDeleteLinkedNode);
        dataDeleted = true;
      }

      if (currentNode.getNext() != null) {
        currentNode = currentNode.getNext();
      }
    } while (currentNode.getNext() != null);

    if (!dataDeleted) {
      throw new NullPointerException();
    }
  }

  public void removeAtIndex(int index) {
    nullListCheck();

    // Remove headNode when index is zero, otherwise remove the referenced Node
    if (index == 0) {
      removeFirst();

    } else {
      Node currentNode = headNode;
      int currentIndex = 0;

      while (currentNode.getNext() != null && currentIndex < index) {
        if (currentIndex == index - 1) {
          // Get the Node to remove (the next Node) and it's reference
          Node nodeToDelete = currentNode.getNext();
          Node nodeToDeleteLinkedNode = nodeToDelete.getNext();

          // Remove the link to any other Node from the Node to remove
          nodeToDelete.setNext(null);

          // Set the removed link to the currentNode
          currentNode.setNext(nodeToDeleteLinkedNode);
          return;
        }

        currentNode = currentNode.getNext();
        currentIndex += 1;
      }

      throw new IndexOutOfBoundsException();
    }
  }

  public void removeLast() {
    nullListCheck();

    // If the list contains one element, delete the list.
    if (headNode.getNext() == null) {
      headNode = null;
      return;
    }

    Node currentNode = headNode;

    // Iterate to find the second-to-last element
    while (currentNode.getNext().getNext() != null) {
      currentNode = currentNode.getNext();
    }

    // Remove the link between the second-to-last element and the last element.
    currentNode.setNext(null);
  }

  public void removeFirst() {
    nullListCheck();

    // Make the headNode the second Node if one exists, otherwise make the headNode null.
    if (headNode.getNext() != null) {
      headNode = headNode.getNext();

    } else {
      headNode = null;
    }
  }

  public void deleteList() {
    if (headNode == null) {
      return;
    }

    Node currentNode = headNode;

    while (currentNode.getNext() != null) {
      Node nextNode = currentNode.getNext();
      currentNode.deleteNode();
      currentNode = nextNode;
    }
  }


  // ------- Get -------

  public Integer getAtIndex(int index) {
    nullListCheck();

    Node currentNode = headNode;
    int currentIndex = 0;

    while (currentNode != null && currentIndex <= index) {
      if (currentIndex == index) {
        return currentNode.getData();
      }

      currentNode = currentNode.getNext();
      currentIndex += 1;
    }

    throw new IndexOutOfBoundsException();
  }

  public Integer getFirst() {
    return headNode.getData();
  }

  public Integer getLast() {
    nullListCheck();

    Node currentNode = headNode;

    while (currentNode.getNext() != null) {
      currentNode = currentNode.getNext();
    }

    return currentNode.getData();
  }


  // ------- Contains -------

  public boolean contains(Integer data) {
    nullListCheck();

    Node currentNode = headNode;

    while (currentNode != null) {
      if (currentNode.getData() == data) {
        return true;
      }

      currentNode = currentNode.getNext();
    }

    return false;
  }


  // - - - - - - - - - - - - - - - Helpers - - - - - - - - - - - - - - -

  private void nullListCheck() {
    if (headNode == null) {
      throw new NullPointerException();
    }
  }

  private void invalidIndexCheck(int index) {
    if (index < 0) {
      throw new IndexOutOfBoundsException();
    }
  }


  // - - - - - - - - - - - - - - - Object callbacks - - - - - - - - - - - - - - -

  @Override
  public String toString() {
    if (headNode != null) {
      return headNode.toString();
    }
    return "empty";
  }

  @Override
  public int hashCode() {
    return super.hashCode();
    //TODO
  }

  @Override
  public boolean equals(Object obj) {
    if (obj instanceof MyLinkedList) {
      // Cast the obj to MyLinkedList
      MyLinkedList other = (MyLinkedList) obj;

      Iterator<Integer> otherIterator = other.iterator();
      Iterator<Integer> thisIterator = this.iterator();

      while (thisIterator.hasNext() && otherIterator.hasNext()) {
        // If the next data in this and the compared are equal, continue iterating. Otherwise, return false.
        if (thisIterator.next().equals(otherIterator.next())) {
          continue;
        }
        return false;
      }

      // Statement reached if this or other cannot be iterated further and have been equal so far.
      return (thisIterator.hasNext() == otherIterator.hasNext());
    }
    return false;
  }


  // - - - - - - - - - - - - - - - Iterable callbacks - - - - - - - - - - - - - - -

  @NotNull
  @Override
  public Iterator<Integer> iterator() {
    return new Iterator<Integer>() {

      Node currentNode = headNode;

      @Override
      public boolean hasNext() {
        return currentNode != null;
      }

      @Override
      public Integer next() {
        // Get the value
        Integer currentData = currentNode.getData();

        // Move to next
        currentNode = currentNode.getNext();

        // Return the value
        return currentData;
      }
    };
  }

  @Override
  public void forEach(Consumer<? super Integer> action) {

  }

  @Override
  public Spliterator<Integer> spliterator() {
    return null;
  }

  // - - - - - - - - - - - - - - - Node class - - - - - - - - - - - - - - -

  private static class Node extends GLinkedListNode<Node, Integer> {

    public Node(Integer data) {
      super(data);
    }
  }
}

File: src/main/kotlin/linkedlist/MyLinkedList.kt
package linkedlist

import java.lang.IndexOutOfBoundsException
import java.lang.NullPointerException

class MyLinkedList() : GLinkedList<Int>, Iterable<Int> {

  // ---------------- Fields & Constructor ----------------

  private var head: Node? = null

  constructor(vararg ints: Int) : this() {
    for (index in (ints.size - 1) downTo 0) {
      addFirst(ints[index])
    }
  }


  // ---------------- Public fun ----------------

  override fun addFirst(data: Int) {
    val newNode = Node(data)

    head?.let { newNode.next = it }
    /*
    Better than:
      if (headNode != null) newNode.linkedNode = headNode
    Same as:
      val tempHead = headNode
      if (tempHead != null) newNode.linkedNode = tempHead
    Explanation:
      In the first method, headNode could become null between the check and assignment of newNode.linkedNode.
     */

    head = newNode
  }

  override fun addLast(data: Int) {
    val newNode = Node(data)

    // If the list is null, make newNode the headNode immediately.
    // Otherwise, iterate to the end of the list and add it there.
    if (head == null) {
      head = newNode

    } else {
      var currentNode = head

      // Iterate to the end of the list, then link newNode to the previous last node
      while (currentNode?.next != null) {
        currentNode = currentNode.next
      }

      currentNode?.next = newNode
    }
  }

  override fun addAtIndex(data: Int, index: Int) {
    if (index < 0) throw IndexOutOfBoundsException()

    // Set up for iteration
    val newNode = Node(data)
    var currentNode = head
    var currentIndex = 0

    // If index is 0, add newNode to start.
    // Otherwise, check if the index exists in the list but is not the last index, and add it there.
    if (index == 0) {
      addFirst(data)
      return

    } else {
      // Loop through this to find index. If found, add and return. Otherwise, add to last index or throw exception.
      while (currentNode != null && currentIndex < index) {
        if (currentIndex == index - 1) {
          newNode.next = currentNode.next
          currentNode.next = newNode
          return
        }
        currentNode = currentNode.next
        currentIndex += 1
      }
    }

    // If the index is the last position, add it there. Otherwise, throw an exception.
    if (currentIndex == index) currentNode?.next = newNode else throw IndexOutOfBoundsException()
  }

  override fun setFirst(data: Int) {
    if (head == null) {
      throw NullPointerException()

    } else {
      val newNode = Node(data)
      newNode.next = head?.next
      head = newNode
    }
  }

  override fun setLast(data: Int) {
    // If the list is null, make newNode the headNode immediately.
    // Otherwise, iterate to the end of the list and add it there.
    if (head == null) {
      throw NullPointerException()

    } else {
      var currentNode = head

      // Iterate to the end and link the last Node to newNode
      while (currentNode?.next != null) currentNode = currentNode.next
      currentNode?.data = data
    }
  }

  override fun setAtIndex(data: Int, index: Int) {
    if (index < 0) throw IndexOutOfBoundsException()
    if (head == null) throw NullPointerException()

    // If index is 0, set headNode to newNode.
    // Otherwise, check if index is in range and add it there.
    if (index == 0) {
      setFirst(data)
      return

    } else {
      var currentNode = head
      var currentIndex = 0

      while (currentNode != null && currentIndex <= index) {
        if (currentIndex == index) {
          currentNode.data = data
          return
        }

        currentNode = currentNode.next
        currentIndex += 1
      }
    }

    // Throw invalid index exception
    throw IndexOutOfBoundsException()
  }

  override fun removeFirst() {
    if (head == null) throw NullPointerException()

    // If the list only contains headNode, delete it
    // Otherwise, make headNode's linkedNode null and make the second Node the new head
    if (head?.next == null) {
      head = null

    } else {
      val newHeadNode = head?.next
      head?.next = null
      head = newHeadNode
    }
  }

  override fun removeLast() {
    if (head == null) throw NullPointerException()

    // If the list only contains headNode, delete it
    // Otherwise, iterate to the end of the list and make the second-to-last Node link null (unlink the last Node)
    if (head?.next == null) {
      head = null

    } else {
      var currentNode = head

      // Iterate until currentNode is the second-to-last Node, then delete it's reference to the last Node
      while (currentNode?.next?.next != null) currentNode = currentNode.next
      currentNode?.next = null
    }
  }

  override fun removeAtIndex(index: Int) {
    if (head == null) throw NullPointerException()

    if (index == 0) {
      removeFirst()
      return
    }

    var currentNode = head
    var currentIndex = 0

    while (currentNode?.next != null && currentIndex < index) {
      if (currentIndex == index - 1) {
        val newRef = currentNode.next?.next
        currentNode.next?.next = null
        currentNode.next = newRef
        return
      }

      currentNode = currentNode.next
      currentIndex += 1
    }
    throw IndexOutOfBoundsException()
  }

  override fun remove(data: Int) {
    if (head == null) throw NullPointerException()

    var currentNode = head

    if (currentNode?.data == data) {
      removeFirst()
      return
    }

    while (currentNode?.next != null) {
      if (currentNode.next?.data == data) {
        val link = currentNode.next?.next
        currentNode.next?.next = null
        currentNode.next = link
        return
      }
      currentNode = currentNode.next
    }
    throw NullPointerException()
  }

  override fun getFirst(): Int {
    val tempNode = head;
    if (tempNode == null) throw NullPointerException() else return tempNode.data
  }

  override fun getLast(): Int {
    if (head == null) throw NullPointerException()

    var currentNode = head
    while (currentNode?.next != null) currentNode = currentNode.next
    if (currentNode?.data != null) return currentNode.data else throw NullPointerException()
  }

  override fun getAtIndex(index: Int): Int {
    if (head == null) throw NullPointerException()
    if (index < 0) throw IndexOutOfBoundsException()

    var currentNode = head
    var currentIndex = 0

    while (currentNode != null && currentIndex <= index) {
      if (currentIndex == index) return currentNode.data
      currentNode = currentNode.next
      currentIndex += 1
    }
    throw IndexOutOfBoundsException()
  }

  override fun contains(data: Int): Boolean {
    if (head == null) throw NullPointerException()

    var currentNode = head

    while (currentNode != null) {
      if (currentNode.data == data) return true
      currentNode = currentNode.next
    }
    return false
  }

  override fun equals(other: Any?): Boolean {
    if (other !is MyLinkedList) return false

    var current = head
    var otherCurrent = other.head

    while (current != null && otherCurrent != null) {
      if (current != otherCurrent) return false
      current = current.next
      otherCurrent = otherCurrent.next
    }
    return current == otherCurrent
  }

  override fun toString(): String {
    if (head != null) {
      return head.toString()
    }
    return "empty"
  }

  override fun hashCode(): Int {
    var result = 1
    var current = head

    while (current != null) {
      result = result * 31 + current.hashCode()
      current = current.next
    }
    return result
  }

  override fun iterator(): Iterator<Int> {
    return object : Iterator<Int> {

      var currentNode = head;

      override fun hasNext(): Boolean {
        return currentNode != null
      }

      override fun next(): Int {
        val returnInt: Int? = currentNode?.data

        if (returnInt == null) {
          throw NullPointerException()

        } else {
          currentNode = currentNode?.next
          return returnInt
        }
      }
    }
  }

  fun isEmpty(): Boolean {
    return head == null
  }

  fun isNotEmpty(): Boolean {
    return head != null
  }


  // ---------------- Node data class ----------------

  private data class Node(var data: Int) {
    var next: Node? = null

    override fun toString(): String {
      return "[" + data + "]" + " -> " + if (next == null) "null" else next.toString()
    }

    override fun hashCode(): Int {
      val nextData = next?.data ?: 0
      return 31 * (data.hashCode() + nextData.hashCode())
    }

    override fun equals(other: Any?): Boolean {
      if (other !is Node) return false

      return other.data == data && other.next == next
    }
  }
}